### ISSUE TITLE
opentelemetry / tracing support - q&a

### ISSUE THREAD

**Issue by mstrYoda** [valyala/fasthttp#940]
Hi, we want to enable open telemetry in fasthttp. Actually we started to write a helper library, somehow we can support tracing for the requests written like below:

```
func getUserFasthttp(ctx context.Context, id string) string {
	req := fasthttp.AcquireRequest()
	res := fasthttp.AcquireResponse()
	defer fasthttp.ReleaseRequest(req)
	defer fasthttp.ReleaseResponse(res)

	req.SetRequestURI("http://10.43.200.175:30425/hello")

	span := tracing.StartSpan(ctx, req, "test")
	err := fasthttp.Do(req, res)
	tracing.EndSpan(span, res.StatusCode(), err)

	if err != nil {
		panic(err)
	}

	return "test"
}
```

but, if developers choose to use fasthttp.Get/Post approach, we could not add telemetry data:

```
func getUserFasthttpWithGet(ctx context.Context, id string) string {
	span := tracing.StartSpan(ctx, nil, "test")
	var responseBody []byte
	status, _, err := fasthttp.Get(responseBody, "http://test/hello")
	tracing.EndSpan(span, status, err)

	if err != nil {
		panic(err)
	}

	return "test"
}
```

Because we could not access to the internal Request/Response struct.

We thought that we can add another method to fasthttp source code with tracing support, or, we can add hooks to thoose methods in below:

```
client.go line 826

func clientGetURL(dst []byte, url string, c clientDoer) (statusCode int, body []byte, err error) {
	req := AcquireRequest()

        // here we can run hooks
        //hook(req)

	statusCode, body, err = doRequestFollowRedirectsBuffer(req, dst, url, c)

	ReleaseRequest(req)
	return statusCode, body, err
}

client.go line 926

func doRequestFollowRedirectsBuffer(req *Request, dst []byte, url string, c clientDoer) (statusCode int, body []byte, err error) {
	resp := AcquireResponse()
	bodyBuf := resp.bodyBuffer()
	resp.keepBodyBuffer = true
	oldBody := bodyBuf.B
	bodyBuf.B = dst

	statusCode, _, err = doRequestFollowRedirects(req, resp, url, defaultMaxRedirectsCount, c)

        // here we can run hooks
        //hook(req)

	body = bodyBuf.B
	bodyBuf.B = oldBody
	resp.keepBodyBuffer = false
	ReleaseResponse(resp)

	return statusCode, body, err
}
```

The hook() functions might be used in several ways for example, telemetry support, request/resp logging, header manipulation etc...

What are your thoughts?

**Comment by previousdeveloper** [comment]
+1

**Comment by erikdubbelboer** [comment]
Where would you suggest to add these hook properties? On `Client/HostClient` or on `Request`?

**Comment by mstrYoda** [comment]
> Where would you suggest to add these hook properties? On `Client/HostClient` or on `Request`?

It needs both request and response, so we need somewhere which we can access both at the same time. I will dive source code and read more carefully, but currently I'm not sure.

Do you have any suggestion about it? I might do an example pr for it.

**Comment by aschenmaker** [comment]
May be you can add context to Request Struct, just like `net/http` request, we can use that context to do traceing

**Comment by gaby** [comment]
Any updates on this?

**Comment by 0angelic0** [comment]
Hi, any update for fasthttp opentelemetry?

**Comment by erikdubbelboer** [comment]
No update, a pull request is still welcome.

**Comment by mstrYoda** [comment]
I want to work on this, can you assign it to me?

**Comment by erikdubbelboer** [comment]
@mstrYoda go got it! 👍

**Comment by asecondo** [comment]
@mstrYoda do you have an idea of when you're aiming to complete this? i'd love to replace some homegrown wrapped tracing instrumentation with officially supported instrumentation 🙂 i'm also happy to co-author a PR if you'd like.

**Comment by mstrYoda** [comment]
> @mstrYoda do you have an idea of when you're aiming to complete this? i'd love to replace some homegrown wrapped tracing instrumentation with officially supported instrumentation 🙂 i'm also happy to co-author a PR if you'd like.

Hi @asecondo with my team we scheduled a pair session to work on this issue (to bring OTEL span propagation to fasthttp, hopefully :) ). We can collaborate on it if you would like to, after creating a structure. I will inform you about it 👍

**Comment by asecondo** [comment]
hey @mstrYoda that sounds good! i have a branch locally with a basic structure that i started on as well. let me know if you want to start with that, or if you have one that we can collaborate on.

**Comment by nithinputhenveettil** [comment]
Just a gentle ping to check on the update here. Are you guys still working on this?

**Comment by MrWildanMD** [comment]
Any update on this?

**Comment by mdenushev** [comment]
@erikdubbelboer Hi, I also writing otel wrapper for client and it is really awkward on client side, because it's hard to propagate span information through many abstraction layers(

What do you think about moving `userValues` field from Ctx to Request struct? It seems reasonable and would help propagate additional context to client side. What do you think?

**Comment by gaby** [comment]
@mdenushev I'd say start with the server, that's the most used part.

**Comment by mdenushev** [comment]
> I'd say start with the server, that's the most used part.

I already covered server part and `RequestCtx` to `context.Context` conversion with span propagation. Client is the only part that missing now (and the most painful).

**Comment by erikdubbelboer** [comment]
@mdenushev can you give an example of how you would use it if `userValues` was moved to `Request`?

**Comment by mdenushev** [comment]
@erikdubbelboer I would use it similar to `net/http.Request` context. Set span to request values, set additional info, also it will be possible to use base Client without custom interfaces for context propagation

**Comment by erikdubbelboer** [comment]
So when you create a Request you don't use it right away but instead pass it around various places and you want to keep some state with it that you don't also want to pass around separately?

**Comment by mdenushev** [comment]
Yes, this is the reason, also now default client interface is `Do(req, resp) err`, so there no way to pass user-defined context to it without wrappers.
